nidx/src/maintenance/scheduler.rs

```diff
@@ -158,23 +169,112 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
+        i64::from(before_seq),
     )
     .fetch_all(&meta.pool)
     .await?;
     for m in merges {
-        if m.len() > 3 {
-            MergeJob::create(meta, &m, oldest_confirmed_seq as i64).await?;
+        if m.segment_ids.len() > 3 {
+            MergeJob::create(meta, m.index_id.into(), &m.segment_ids, before_seq).await?;
         }
     }
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod merge_scheduling {
+        use std::collections::HashSet;
+
+        use nidx_types::Seq;
+        use uuid::Uuid;
+
+        use super::*;
+
+        use crate::metadata::{Index, IndexId, IndexKind, NidxMetadata, Shard};
+
+        #[sqlx::test]
+        async fn test_schedule_merges_for_shard_with_single_index(pool: sqlx::PgPool) {
+            let meta = NidxMetadata::new_with_pool(pool).await.unwrap();
+            let kbid = Uuid::new_v4();
+            let shard = Shard::create(&meta.pool, kbid).await.unwrap();
+            let index = Index::create(&meta.pool, shard.id, IndexKind::Vector, Some("multilingual")).await.unwrap();
+            let mut seq: i64 = 0;
+
+            for _ in 0..10 {
+                let segment = Segment::create(&meta.pool, index.id, Seq::from(seq)).await.unwrap();
+                segment.mark_ready(&meta.pool, 50, 1000).await.unwrap();
+                seq += 1;
+            }
+
+            // creation of shards/indexes/segments don't trigger any merge job
+            assert!(MergeJob::take(&meta.pool).await.unwrap().is_none());
+
+            schedule_merges(&meta, Seq::from(100)).await.unwrap();
+
+            // one job has been scheduled for the index
+            let mut jobs = vec![];
+            while let Some(job) = MergeJob::take(&meta.pool).await.unwrap() {
+                jobs.push(job)
+            }
+            assert_eq!(jobs.len(), 1);
+            assert_eq!(IndexId::from(jobs[0].index_id), index.id);
+        }
+
+        #[sqlx::test]
+        async fn test_schedule_merges_for_shard_with_multiple_indexes(pool: sqlx::PgPool) {
+            let meta = NidxMetadata::new_with_pool(pool).await.unwrap();
+            let kbid = Uuid::new_v4();
+            let shard = Shard::create(&meta.pool, kbid).await.unwrap();
+
+            let indexes = vec![
+                Index::create(&meta.pool, shard.id, IndexKind::Vector, Some("multilingual")).await.unwrap(),
+                Index::create(&meta.pool, shard.id, IndexKind::Vector, Some("english")).await.unwrap(),
+                Index::create(&meta.pool, shard.id, IndexKind::Text, Some("fulltext")).await.unwrap(),
+                Index::create(&meta.pool, shard.id, IndexKind::Paragraph, Some("keyword")).await.unwrap(),
+                Index::create(&meta.pool, shard.id, IndexKind::Relation, Some("relation")).await.unwrap(),
+            ];
+            let mut seq: i64 = 0;
+
+            for index in indexes.iter() {
+                for _ in 0..10 {
```

Comment:
nitpick: I would flip the order of these two loops to be a bit closer to reality, i.e: loop over resources first and loop over indexes inside.